certora/specs/Reverts.spec

```diff
@@ -0,0 +1,313 @@
+    MorphoHarness.Id id = Morpho.libId(marketParams);
+
+    bool hasCuratorRole = hasCuratorRole(e.msg.sender);
+    address asset = asset();
+    uint256 lastUpdate = Morpho.lastUpdate(id);
+    uint256 pendingCapValidAt;
+    _, pendingCapValidAt = pendingCap(id);
+    uint184 supplyCap;
+    uint64 removableAt;
+    supplyCap, _, removableAt = config(id);
+
+    submitCap@withrevert(e, marketParams, newSupplyCap);
+
+    assert e.msg.value != 0 ||
+           !hasCuratorRole ||
+           marketParams.loanToken != asset ||
+           lastUpdate == 0 ||
+           pendingCapValidAt != 0 ||
+           removableAt != 0 ||
+           newSupplyCap == assert_uint256(supplyCap)
+        => lastReverted;
+}
+
+rule submitMarketRemovalRevertCondition(env e, MetaMorphoHarness.MarketParams marketParams) {
+    MorphoHarness.Id id = Morpho.libId(marketParams);
+
+    bool hasCuratorRole = hasCuratorRole(e.msg.sender);
+    uint256 pendingCapValidAt;
+    _, pendingCapValidAt = pendingCap(id);
+    uint184 supplyCap;
+    bool enabled;
+    uint64 oldRemovableAt;
+    supplyCap, enabled, oldRemovableAt = config(id);
+
+    // Safe require because it is a verified invariant.
+    require isTimelockInRange();
+    // Safe require as it corresponds to year 2262.
+    require e.block.timestamp < 2^63;
+
+    submitMarketRemoval@withrevert(e, marketParams);
+
+    assert lastReverted <=>
+        e.msg.value != 0 ||
+        !hasCuratorRole ||
+        pendingCapValidAt != 0 ||
+        supplyCap != 0 ||
+        !enabled ||
+        oldRemovableAt != 0;
+}
+
+rule setSupplyQueueInputValidation(env e, MorphoHarness.Id[] newSupplyQueue) {
+    bool hasAllocatorRole = hasAllocatorRole(e.msg.sender);
+    uint256 maxQueueLength = maxQueueLength();
+    uint256 i;
+    require i < newSupplyQueue.length;
+    uint184 anyCap;
+    anyCap, _, _ = config(newSupplyQueue[i]);
+
+    setSupplyQueue@withrevert(e, newSupplyQueue);
+
+    assert e.msg.value != 0 ||
+           !hasAllocatorRole ||
+           newSupplyQueue.length > maxQueueLength ||
+           anyCap == 0
+        => lastReverted;
+}
+
+rule updateWithdrawQueueInputValidation(env e, uint256[] indexes) {
+    bool hasAllocatorRole = hasAllocatorRole(e.msg.sender);
+    uint256 i;
+    require i < indexes.length;
+    uint256 j;
+    require j < indexes.length;
+    uint256 anyIndex = indexes[i];
+    uint256 oldLength = withdrawQueueLength();
+    uint256 anyOtherIndex = indexes[j];
+
+    updateWithdrawQueue@withrevert(e, indexes);
+
+    assert e.msg.value != 0 ||
+           !hasAllocatorRole ||
+           anyIndex > oldLength ||
+           (i != j && anyOtherIndex == anyIndex)
+        => lastReverted;
+}
+
+rule reallocateInputValidation(env e, MetaMorphoHarness.MarketAllocation[] allocations) {
+    bool hasAllocatorRole = hasAllocatorRole(e.msg.sender);
+
+    reallocate@withrevert(e, allocations);
+
+    assert e.msg.value != 0 ||
+           !hasAllocatorRole
+        => lastReverted;
```

Comment:
- also reverts if a given market isn't enabled
- could we check that it reverts if supplied!=withdrawn

Reply:
Added in #333